src/coreclr/gc/gc.cpp

```diff
@@ -48926,10 +48941,9 @@ GCHeap::Alloc(gc_alloc_context* context, size_t size, uint32_t flags REQD_ALIGN_
 #ifdef MULTIPLE_HEAPS
         if (flags & GC_ALLOC_FINALIZE)
         {
-            // the heap may have changed due to heap balancing - it's important
+            // the heap may have changed due to heap balancing or heaps going out of service
             // to register the object for finalization on the heap it was allocated on
-            hp = acontext->get_alloc_heap()->pGenGCHeap;
-            assert ((newAlloc == nullptr) || (hp == gc_heap::heap_of ((uint8_t*)newAlloc)));
+            hp = (newAlloc == nullptr) ? acontext->get_alloc_heap()->pGenGCHeap : gc_heap::heap_of ((uint8_t*)newAlloc);
```

Comment:
looks like you've simply gotten rid of the assert. probably meant something like
```cplusplus
            // the heap may have changed due to heap balancing - it's important
            // to register the object for finalization on the heap it was allocated on
            // but if that heap went out of service it wouldn't be the same 
            hp = acontext->get_alloc_heap()->pGenGCHeap;
            assert ((newAlloc == nullptr)
#ifndef DYNAMIC_HEAP_COUNT
                        || (hp == gc_heap::heap_of ((uint8_t*)newAlloc))
#endif //!DYNAMIC_HEAP_COUNT
             );
```

Reply:
I have observed the assert firing, and I think it was because `acontext->get_alloc_heap()->pGenGCHeap` went out of service. It wouldn't be good to try to register an object for finalization on that heap.

Thus I think this part: `hp = ... gc_heap::heap_of ((uint8_t*)newAlloc)` is actually important.

I have #ifdef'd the code so we use the new version (without the assert) in the DYNAMIC_HEAP_COUNT case and the previous code otherwise.

Reply:
ahh indeed we need to set the correct hp. I did not notice we have the finalization part below this code.